### PR TITLE
[MXNET-1185] [WIP] Support large array in several operators

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
This PR fixed the large array issue (https://github.com/apache/incubator-mxnet/issues/13036, https://github.com/apache/incubator-mxnet/issues/13070) in the following operators:
ndarray.ones
ndarray.zeros
ndarray.sum
ndarray.arange
ndarray.slice
ndarray.random.uniform
ndarray.empty

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support large arrays (> 3billion elements) in certain operators
- [X] nightly tests on arrays of 5 billion elements
## Comments ##
- More PRs in the future to support other operators
- Need a thorough performance test
- Rely on a defined data type index_t which is int64_t for now. It can be tuned based on various platforms.

### CODE REVIEW

File: python/mxnet/base.py [hunk 1]
```diff
@@ -215,6 +215,7 @@ def _load_lib():
 # type definitions
 mx_uint = ctypes.c_uint
 mx_float = ctypes.c_float
+mx_long = ctypes.c_longlong
```
Comment (zheng-da) [inline python/mxnet/base.py]: should this be longlong or just long?

**Comment by anirudhacharya** [comment]
@mxnet-label-bot [pr-awaiting-review]

**Comment by apeforest** [comment]
@yuxihu That's a good question. We have not yet run a full performance test. @wkcn did some initial test and did not see much performance variation between int32_t and int64_t: https://github.com/wkcn/c_performance

**Comment by wkcn** [comment]
@apeforest Could you please add the test? https://github.com/apache/incubator-mxnet/issues/13070
Thank you!

**Comment by apeforest** [comment]
> @apeforest Could you please add the test? #13070
> Thank you!

Added: test_ndarray_empty()

File: scala-package/native/src/main/native/org_apache_mxnet_native_c_api.cc [hunk 2]
```diff
@@ -1519,14 +1519,14 @@ JNIEXPORT jint JNICALL Java_org_apache_mxnet_LibInfo_mxSymbolCreateFromFile
 
 int FillSymbolInferShape
   (JNIEnv *env, jmethodID listAppend, jobject joutData,
-    mx_uint shapeSize, const mx_uint *shapeNdim, const mx_uint **shapeData) {
+    mx_uint shapeSize, const mx_uint *shapeNdim, const dim_t **shapeData) {
   for (size_t i = 0; i < shapeSize; ++i) {
     jintArray jshape = env->NewIntArray(shapeNdim[i]);
```
[the remark this answers is not in the document]
Comment (apeforest) [reply scala-package/native/src/main/native/org_apache_mxnet_native_c_api.cc]: @andrewfayres Thanks for your review. I have updated code based on your suggestion.

File: scala-package/native/src/main/native/org_apache_mxnet_native_c_api.cc [hunk 3]
```diff
@@ -1578,7 +1578,7 @@ JNIEXPORT jint JNICALL Java_org_apache_mxnet_LibInfo_mxSymbolInferShape
                                &auxShapeNdim,
                                &auxShapeData,
                                &complete);
-  env->ReleaseIntArrayElements(jargShapeData, argShapeData, 0);
+  env->ReleaseLongArrayElements(jargShapeData, argShapeData, 0);
```
[the remark this answers is not in the document]
Comment (apeforest) [reply scala-package/native/src/main/native/org_apache_mxnet_native_c_api.cc]: updated.

**Comment by lanking520** [comment]
@apeforest Regarding what you have changed the JNI on Scala, I suggest to apply some changes to the following line as well:
1. https://github.com/apache/incubator-mxnet/blob/master/scala-package/core/src/main/scala/org/apache/mxnet/LibInfo.scala#L43 `Array[Long]`

2. What is `dim_t`, is there any Java type I can change correspondingly? https://github.com/apache/incubator-mxnet/blob/master/scala-package/core/src/main/scala/org/apache/mxnet/LibInfo.scala#L78

3. https://github.com/apache/incubator-mxnet/blob/master/scala-package/core/src/main/scala/org/apache/mxnet/LibInfo.scala#L239

@andrewfayres WDYT?

File: R-package/src/base.h [hunk 4]
```diff
@@ -354,8 +354,8 @@ inline std::vector<std::string> SafeGetListNames(const Rcpp::List& src) {
  * \param rshape The dimension in R
  * \return A internal vector representation of shapes in mxnet.
  */
-inline std::vector<mx_uint> Dim2InternalShape(const Rcpp::Dimension &rshape) {
-  std::vector<mx_uint> shape(rshape.size());
+inline std::vector<dim_t> Dim2InternalShape(const Rcpp::Dimension &rshape) {
```
Comment (yuxihu) [inline R-package/src/base.h]: QQ: what is the type of mx_uint and dim_t?

File: src/operator/mxnet_op.h [hunk 5]
```diff
@@ -536,22 +536,22 @@ struct Kernel<OP, cpu> {
...
   template<typename PRIMITIVE_OP, typename DType, typename ...Args>
-  static void LaunchTuned(mshadow::Stream<cpu> *, const int N, Args... args) {
+  static void LaunchTuned(mshadow::Stream<cpu> *, const size_t N, Args... args) {
 #ifdef _OPENMP
     const int omp_threads = engine::OpenMP::Get()->GetRecommendedOMPThreadCount();
     if (omp_threads < 2 || !tuned_op<PRIMITIVE_OP, DType>::UseOMP(
-      static_cast<size_t>(N), static_cast<size_t>(omp_threads))) {
-      for (int i = 0; i < N; ++i) {
+      N, static_cast<size_t>(omp_threads))) {
+      for (size_t i = 0; i < N; ++i) {
```
Comment (yuxihu) [inline src/operator/mxnet_op.h]: why size_t here and index_t on L549?

**Comment by apeforest** [comment]
@lanking520 @andrewfayres After offline discussions, I have updated the PR by limiting the changes in scala package to its JNI interface and leave a TODO item to systematically update Scala code to support large array. Please help to review the code again since there are still a few scala-unit test failure. Thanks!

**Comment by apeforest** [comment]
Change this PR to [WIP] because we want to merge a smaller part first (https://github.com/apache/incubator-mxnet/pull/13418)

**Comment by apeforest** [comment]
@mxnet-label-bot update [pr-work-in-progress]

File: src/operator/mxnet_op.h (hunk 5 above)
Comment (apeforest) [reply src/operator/mxnet_op.h]: Because omp parallel block requires index to be signed integer, while size_t is unsigned.

**Comment by roywei** [comment]
@apeforest do you plan to still use this PR for part 2? If not maybe close it first.
Thanks!

**Comment by apeforest** [comment]
@roywei We can close this PR for now. I will reopen it when it comes to all the language binding support. Thanks!

**Pull request closed without merging**